Add support for cookie file to media_extractor

RickeyWard opened the pull request:
## Proposed change
media_extractor component: add support for a cookie-file

the library media_extractor uses supports loading a cookie file but doesn't expose any way to use it. This small PR checks for existence of a txt file "media_extractor_cookies.txt" in your config directory and if it exists loads it. This file is not altered by the library and can be read only. The file follows the cookie_spec for Netscape HTTP Cookie File (can be generated with browser extensions like opencookies, manually, etc)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

The motivation behind this change is that some services provide a better experience when logged in. For example if you have twitch turbo, or are subscribed to a twitch streamer logging in will prevent the video stream from showing an "an ad is currently playing come back after the break" message until the ads are over as well as if a youtube video is private or purchased you can only watch it while logged in.

- This PR fixes or closes issue: fixes NONE
- This PR is related to issue:  NONE
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30152

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
// its not clear to me how a test could be written that would benefit the change as it is

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
https://github.com/home-assistant/home-assistant.io/pull/30129
If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:
(as a first time contributor I do not believe I'm confident enough in the code base or the the culture to review PRs. I'm checking this box because I made a good faith effort to look through the open PRs for one that I felt comfortable reviewing)
- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

RickeyWard committed:
media_extractor comp -> add support for cookie file

RickeyWard committed:
Update __init__.py

RickeyWard committed:
Update __init__.py

RickeyWard committed:
Merge branch 'dev' into dev

joostlek reviewed (commented):
What would you use this for?

joostlek commented on homeassistant/components/media_extractor/__init__.py:
```diff
@@ -106,7 +108,19 @@ def extract_and_send(self) -> None:
+        )
+        if os.path.isfile(cookies_file):
+            ydl = YoutubeDL(
+                {"quiet": True, "logger": _LOGGER, "cookiefile": cookies_file}
```
Can we maybe just update the options dict we pass in if its a file instead of duplicating the creation of the object

joostlek commented on homeassistant/components/media_extractor/__init__.py:
```diff
@@ -106,7 +108,19 @@ def extract_and_send(self) -> None:
+            ydl = YoutubeDL(
+                {"quiet": True, "logger": _LOGGER, "cookiefile": cookies_file}
+            )
+            _LOGGER.info("Media extractor loaded cookies file from: %s", cookies_file)
```
Please don't use info, debug would be more fitting here

RickeyWard replied:
no problem, fixed.

joostlek reviewed (changes requested):
I got 2 remarks. I will have to verify this change with some others because I haven't worked a lot with integrations doing disk IO. I would like to see a test, but I am also unsure on how to approach that. Probably by trying to create the file and then mocking YoutubeDL and check what options dict we got as input.

RickeyWard committed:
fixed pr request, added test

joostlek reviewed (changes requested):
Just discussed with someone, can we put the file in `<config_dir>/media_extractor/cookies.txt`?

RickeyWard committed:
update test with valid cookie file

joostlek commented:
And when its ready for review, don't forget to press the button

RickeyWard committed:
move cookies to subdirectory

joostlek commented on homeassistant/components/media_extractor/__init__.py:
```diff
@@ -106,7 +108,18 @@ def extract_and_send(self) -> None:
     def get_stream_selector(self) -> Callable[[str], str]:
         """Return format selector for the media URL."""
-        ydl = YoutubeDL({"quiet": True, "logger": _LOGGER})
+        cookies_file = os.path.join(
+            self.hass.config.config_dir, "media_extractor", "cookies.txt"
+        )
```
I got the advice to use path lib for this, I am currently in the train so can't really check https://docs.python.org/3/library/pathlib.html

RickeyWard replied:
I don't see any advantage in this but if that's the prefered way I'll use it. It means creating a complex type that has to be converted back to a string to be used.

RickeyWard committed:
use pathlib instead of os.path

RickeyWard commented:
I added a test that creates the file and then verifies the logger sees the correct log message if the file exists. I'm not sure if its the most useful of tests, but it does show the correct path is taken if the file exists.

> Just discussed with someone, can we put the file in `<config_dir>/media_extractor/cookies.txt`?

moved to `media_extractor/cookies.txt`

> I got the advice to use path lib for this

converted to use pathlib

RickeyWard committed:
Merge branch 'dev' into dev

RickeyWard committed:
Merge branch 'dev' into dev

RickeyWard committed:
Merge branch 'dev' into dev

RickeyWard committed:
Merge branch 'dev' into dev

RickeyWard committed:
Merge branch 'dev' into dev

RickeyWard commented:
Is there anything else I need to do to get this merged?

joostlek commented:
Oh lol, I approved it, waited for the CI to finish and then I forgot to merge 🤦‍♂️
